docs/source/guide/migrations3.rst

```diff
@@ -0,0 +1,160 @@
+.. _guide_migration_s3:
+
+Amazon S3
+=========
+Boto 2.x contains a number of customizations to make working with Amazon S3 buckets and keys easy.
+
+Creating the Connection
+-----------------------
+Boto 3 has both low-level clients and higher-level resources. For Amazon S3, the higher-level resources are the most similar to Boto 2.x's ``s3`` module::
+
+    # Boto 2.x
+    import boto
+    s3_connection = boto.connect_s3()
+
+    # Boto 3
+    import boto3
+    s3 = boto3.resource('s3')
+
+Creating a Bucket
+-----------------
+Creating a bucket in Boto 2 and Boto 3 is very similar, except that in Boto 3 all action parameters must be passed via keyword arguments and a bucket configuration must be specified manually::
+
+    # Boto 2.x
+    s3_connection.create_bucket('mybucket')
+    s3_connection.create_bucket('mybucket', location=Location.USWest)
+
+    # Boto 3
+    s3.create_bucket(Bucket='mybucket')
+    s3.create_bucket(Bucket='mybucket', CreateBucketConfiguration={
+        'Location': 'us-west-1'})
+
+Storing Data
+------------
+Storing data from a file, stream, or string is easy::
+
+    # Boto 2.x
+    from boto.s3.key import Key
+    key = Key('hello.txt')
+    key.set_contents_from_file('/tmp/hello.txt')
+
+    # Boto 3
+    s3.Key('mybucket', 'hello.txt').put(Body=open('/tmp/hello.txt'))
+
+
+Accessing a Bucket
+------------------
+Getting a bucket is easy with Boto 3's resources, however these do not automatically validate whether a bucket exists::
+
+    # Boto 2.x
+    bucket = s3_connection.get_bucket('mybucket', validate=False)
+    exists = s3_connection.lookup('mybucket')
+
+    # Boto 3
+    bucket = s3.Bucket('mybucket')
+    bucket.wait_until_exists(max_attempts=2, service_args={'Foo': 1})
+    try:
+        s3.meta.client.head_bucket(Bucket='mybucket')
+        exists = True
+    except ClientError:
```

Comment:
This is not necessarily true you can get a 403 Access Denied in which the bucket does exist but you do not have HEAD permissions.

The exists() function for resources is starting to look like more and more like something that we would want to implement.